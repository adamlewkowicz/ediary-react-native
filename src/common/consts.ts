export const PRODUCT_UNITS = ['g', 'mg', 'kg', 'ml', 'l'] as const;
export const BASE_MACRO_ELEMENTS = ['carbs', 'prots', 'fats'] as const
export const MACRO_ELEMENTS = [...BASE_MACRO_ELEMENTS, 'kcal'] as const;