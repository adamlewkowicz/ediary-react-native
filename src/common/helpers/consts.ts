import { MacroElements } from '../../types';
import { DiaryTemplate } from '../../store/reducers/diary';

export const baseMacro: MacroElements = {
  carbs: 0,
  prots: 0,
  fats: 0,
  kcal: 0
}

export const defaultTemplates = [
  {
    id: 1,
    name: 'Śniadanie I',
    time: '08:30:00',
  },
  {
    id: 2,
    name: 'Śniadanie II',
    time: '11:00:00',
  },
  {
    id: 3,
    name: 'Obiad',
    time: '13:00:00',
  },
  {
    id: 4,
    name: 'Podwieczorek',
    time: '17:00:00',
  },
  {
    id: 5,
    name: 'Kolacja',
    time: '19:00:00',
  },
] as any as DiaryTemplate[];