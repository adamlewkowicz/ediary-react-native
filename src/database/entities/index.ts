export * from './Product';
export * from './Meal';
export * from './MealProduct';
export * from './User';
export * from './ProductPortion';
export * from './Profile';
export * from './ProductImage';
export * from './RunningTraining';
export * from './RunningCoords';