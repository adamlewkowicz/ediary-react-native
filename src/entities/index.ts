export * from './Product';
export * from './Meal';
export * from './MealProduct';