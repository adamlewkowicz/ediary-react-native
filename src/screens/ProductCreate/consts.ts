
export const NUTRITION_INPUTS = [
  {
    title: 'Węglowodany',
    property: 'carbs',
    nextRef: 'prots'
  },
  {
    title: 'Białka',
    property: 'prots',
    nextRef: 'fats'
  },
  {
    title: 'Tłuszcze',
    property: 'fats',
    nextRef: 'kcal'
  },
  {
    title: 'Kalorie',
    property: 'kcal',
    nextRef: 'barcode'
  }
] as const;

export enum PORTION_TITLE {
  '100g' = 'Opakowanie zawiera',
  'portion' = 'Porcja zawiera',
  'package' = 'Opakowanie zawiera'
};