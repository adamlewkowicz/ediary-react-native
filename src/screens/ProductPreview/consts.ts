
export const PORTIONS = [
  {
    value: 0.25,
    title: '1/4',
  },
  {
    value: 0.5,
    title: '1/2',
  },
  {
    value: 0.75,
    title: '3/4',
  },
  {
    value: 1,
    title: '1',
  },
  {
    value: 1.25,
    title: '1,25',
  },
  {
    value: 1.5,
    title: '1.5',
  },
  {
    value: 1.75,
    title: '1.75',
  },
  {
    value: 2,
    title: '2',
  },
  {
    value: 3,
    title: '3',
  },
  {
    value: 4,
    title: '4',
  },
  {
    value: 5,
    title: '5',
  },
] as const;