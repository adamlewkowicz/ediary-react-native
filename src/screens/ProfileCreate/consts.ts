
export const STEP_TITLES = {
  '0': 'Wybierz płeć',
  '1': 'Twoje pomiary',
  '2': 'Wybierz cel',
};

export const ICON_SIZE = 45;