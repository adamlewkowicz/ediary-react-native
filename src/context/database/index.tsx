import React, { ReactNode } from 'react';
import { createConnection, getRepository, Connection, Repository, ConnectionManager, getConnection } from 'typeorm/browser';
import * as entities from '../../entities';
import { Text } from 'react-native';

export const DatabaseContext = React.createContext<DatabaseContext>({} as any);

interface DatabaseProviderProps {
  children: ReactNode 
}
interface DatabaseProviderState {
  isLoading: boolean
  connection: null | Connection
  repositories: null | Repositories 
}
export class DatabaseProvider extends React.Component<DatabaseProviderProps, DatabaseProviderState> {

  state = {
    connection: null,
    repositories: null,
    isLoading: true
  }

  componentDidMount() {
    this.setupDatabase();
  }

  async setupDatabase() {
    const connection = await createConnection({
      type: 'react-native',
      database: 'test',
      location: 'default',
      logging: ['error', 'query', 'schema'],
      synchronize: false,
      entities: Object.values(entities)
    });

    this.setState({
      connection,
      isLoading: false
    });
  }

  render() {
    
    if (this.state.isLoading) {
      return <Text>≈Åadowanie bazy danych...</Text>;
    }

    return (
      <DatabaseContext.Provider value={this.state as any}>
        {this.props.children}
      </DatabaseContext.Provider>
    );
  }
}

type Repositories = {
  [key in keyof typeof entities]: Repository<typeof entities>
}

interface DatabaseContext extends DatabaseProviderState {
  connection: Connection
  repositories: Repositories
}