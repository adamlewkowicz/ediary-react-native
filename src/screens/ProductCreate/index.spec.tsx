import React from 'react';
import { fireEvent, wait, waitForElementToBeRemoved } from '@testing-library/react-native';
import { Product } from '../../database/entities';
import { ProductCreateScreen } from '.';
import { renderSetup } from '../../../test-utils';
import { APP_ROUTE } from '../../navigation/consts';

describe('<ProductCreateScreen />', () => {

  it('creating new product should work 🥝', async () => {
    const productMock = {
      name: 'Orange Juice',
      brand: 'Bart',
      producer: 'Gardens',
      quantity: 140,
      barcode: '391287145',
      macro: {
        carbs: 249,
        sugars: 41,
        prots: 25,
        fats: 12,
        kcal: 319,
        fattyAcids: 12,
      },
    }
    const productSaveSpy = jest.spyOn(Product, 'save');
    const ctx = renderSetup(<ProductCreateScreen />);

    const nameInput = ctx.getByLabelText('Nazwa');
    const producerInput = ctx.getByLabelText('Producent');
    const brandInput = ctx.getByLabelText('Marka');
    const portionQuantityInput = ctx.getByLabelText('Ilość g w jednej porcji');
    const carbsInput = ctx.getByLabelText('Węglowodany');
    const sugarsInput = ctx.getByLabelText('w tym cukry');
    const protsInput = ctx.getByLabelText('Białko');
    const fatsInput = ctx.getByLabelText('Tłuszcze');
    const fattyAcidsInput = ctx.getByLabelText('w tym kwasy tłuszczowe');
    const kcalInput = ctx.getByLabelText('Kalorie');
    const barcodeInput = ctx.getByLabelText('Kod kreskowy');
    const saveProductButton = ctx.getByLabelText('Zapisz produkt');
  
    fireEvent.changeText(nameInput, productMock.name);
    fireEvent.changeText(brandInput, productMock.brand);
    fireEvent.changeText(producerInput, productMock.producer);
    fireEvent.changeText(portionQuantityInput, String(productMock.quantity));
    fireEvent.changeText(carbsInput, String(productMock.macro.carbs));
    fireEvent.changeText(sugarsInput, String(productMock.macro.sugars));
    fireEvent.changeText(protsInput, String(productMock.macro.prots));
    fireEvent.changeText(fatsInput, String(productMock.macro.fats));
    fireEvent.changeText(fattyAcidsInput, String(productMock.macro.fattyAcids));
    fireEvent.changeText(kcalInput, String(productMock.macro.kcal));
    fireEvent.changeText(barcodeInput, String(productMock.barcode));
    fireEvent.press(saveProductButton);

    await ctx.findByLabelText('Trwa ładowanie');
    await waitForElementToBeRemoved(() => ctx.getByLabelText('Trwa ładowanie'));

    const productPayload = productSaveSpy.mock.calls[0][0];

    expect(productSaveSpy).toHaveBeenCalledTimes(1);
    expect(productPayload).toMatchSnapshot();
    expect(saveProductButton).not.toBeDisabled();
  });

  describe('when presses calories calculate button', () => {

    it('should calculate calories based on provided macro nutriements 🧮', async () => {
      const kcalMock = '0';
      const ctx = renderSetup(<ProductCreateScreen />);

      const kcalInput = ctx.getByLabelText('Kalorie');
      fireEvent.changeText(kcalInput, kcalMock);

      const carbsInput = ctx.getByLabelText('Węglowodany');
      fireEvent.changeText(carbsInput, '100');

      const calculateCaloriesButton = ctx.getByLabelText('Oblicz kalorie');
      fireEvent.press(calculateCaloriesButton);

      await wait(() => 
        expect(kcalInput.getProp('value')).not.toEqual(kcalMock)
      );
    });
    
  });

  describe('when presses barcode scan button', () => {

    it('should navigate to barcode scan screen 🧭', () => {
      const ctx = renderSetup(<ProductCreateScreen />); 

      const barcodeScanButton = ctx.getByLabelText('Zeskanuj');
      fireEvent.press(barcodeScanButton);

      expect(ctx.mocks.navigationContext.navigate).toHaveBeenCalledTimes(1);
      expect(ctx.mocks.navigationContext.navigate).toHaveBeenCalledWith(
        APP_ROUTE.BarcodeScan,
        expect.any(Object)
      );
    });

  });

});