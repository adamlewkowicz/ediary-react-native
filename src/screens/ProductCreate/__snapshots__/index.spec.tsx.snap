// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductCreateScreen /> creating new product should work ü•ù 1`] = `
Object {
  "barcode": "391287145",
  "brand": "Bart",
  "carbs": "249",
  "fats": "12",
  "kcal": "319",
  "macro": Object {
    "carbs": 249,
    "fats": 12,
    "kcal": 319,
    "prots": 25,
  },
  "name": "Orange Juice",
  "portions": Array [
    Object {
      "type": "portion",
      "value": 140,
    },
  ],
  "producer": "Gardens",
  "prots": "25",
  "userId": 0,
}
`;
